lsp-types/src/Language/LSP/MetaModel/CodeGen.hs

```diff
@@ -0,0 +1,911 @@
+  jsonDs <- lift
+    [d|
+     instance J.ToJSON $(TH.conT tn) where
+       toJSON e = J.toJSON ($(TH.varE toBaseTypeN) e)
+
+     instance J.FromJSON $(TH.conT tn) where
+       parseJSON val = do
+         v <- J.parseJSON val
+         $(
+           if custom
+           then [| pure $ $(TH.varE fromBaseTypeN) v |]
+           else [| case $(TH.varE fromBaseTypeN) v of { Just e -> pure e; Nothing -> fail $ "unrecognized enum value for " ++ enumNameString ++ ":" ++ show v; } |]
+          )
+      |]
+
+  pure $ [datad] ++ toBaseTypeD ++ fromBaseTypeD ++ jsonDs
+
+-- | Generate a type declaration corresponding to a type alias.
+genAlias :: TypeAlias -> CodeGenM TH.Dec
+genAlias TypeAlias{name, type_, documentation, since, proposed} = do
+  st <- asks symbolTable
+  tn <- case Map.lookup name st of
+    Just n  -> pure n
+    Nothing -> fail $ "no name for " ++ show name
+  addDocumentation tn documentation since proposed
+
+  rhs <- convertType type_
+
+  ctor <- makeConstrName Nothing name
+  -- We don't actually need JSONKey instances for all the base types, but it's close enough
+  let jsonKey = case type_ of { BaseType _ -> True; _ -> False }
+      newtypeDeriving = TH.DerivClause (Just TH.NewtypeStrategy)
+          ([TH.ConT ''J.ToJSON, TH.ConT ''J.FromJSON] ++ if jsonKey then [TH.ConT ''J.ToJSONKey, TH.ConT ''J.FromJSONKey] else [])
+  pure $ TH.NewtypeD [] tn [] Nothing (TH.NormalC ctor [(noBang, rhs)]) [stockDeriving, newtypeDeriving]
+  --pure $ TH.TySynD tn [] rhs
+
+(.=?) :: (J.KeyValue kv, J.ToJSON v) => J.Key -> Maybe v -> [kv]
+k .=? v = case v of
+  Just v' -> [k J..= v']
+  Nothing -> mempty
+
+-- | Generate a type declaration corresponding to a top-level named struct.
+genStruct :: Structure -> CodeGenM [TH.Dec]
+genStruct s@Structure{name, documentation, since, proposed} = do
+  let structName = name
+
+  st <- asks symbolTable
+  tn <- case Map.lookup name st of
+    Just n  -> pure n
+    Nothing -> fail $ "no name for " ++ show name
+  addDocumentation tn documentation since proposed
+
+  ctor <- lift $ makeToplevelName name
+
+  props <- getStructProperties s
+  args <- for props $ \Property{name, type_, optional, documentation, since, proposed} -> do
+      pty <- convertType type_
+      let mty = case optional of
+            Just True -> TH.ConT ''Maybe `TH.AppT` pty
+            _         -> pty
+      n <- makeFieldName name
+      addDocumentation n documentation since proposed
+      pure (name, (n, noBang, mty))
+
+  let datad = TH.DataD [] tn [] Nothing [TH.RecC ctor (fmap snd args)] [stockDeriving]
+
+  toJsonD <- do
+    (unzip -> (args, pairEs)) <- for props $ \Property{name, optional} -> do
+      n <- lift $ TH.newName "arg"
+      pairE <- case optional of
+        Just True -> lift [| $(TH.litE $ TH.stringL $ T.unpack name) .=? $(TH.varE n) |]
+        _ -> lift [| [$(TH.litE $ TH.stringL $ T.unpack name) J..= $(TH.varE n)] |]
```

Comment:
The spec defining "optional" to be False if it's omitted is confusing and weird... but it's in the spec so